modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs

```diff
@@ -111,62 +113,80 @@ private void _ExportBeginImpl(string[] features, bool isDebug, string path, long
-
-            Directory.CreateDirectory(outputDataDir);
-
-            foreach (string dir in Directory.GetDirectories(publishOutputTempDir, "*", SearchOption.AllDirectories))
+            else if (features.Contains("universal"))
             {
-                Directory.CreateDirectory(Path.Combine(outputDataDir, dir.Substring(publishOutputTempDir.Length + 1)));
+                if (platform == OS.Platforms.MacOS)
+                {
+                    archs.Add("x86_64");
+                    archs.Add("arm64");
+                }
             }
 
-            foreach (string file in Directory.GetFiles(publishOutputTempDir, "*", SearchOption.AllDirectories))
+            foreach (var arch in archs)
             {
-                File.Copy(file, Path.Combine(outputDataDir, file.Substring(publishOutputTempDir.Length + 1)));
+                string ridOS = DetermineRuntimeIdentifierOS(platform);
+                string ridArch = DetermineRuntimeIdentifierArch(arch);
+                string runtimeIdentifier = $"{ridOS}-{ridArch}";
+
+                // Create temporary publish output directory
+
+                string publishOutputTempDir = Path.Combine(Path.GetTempPath(), "godot-publish-dotnet",
+                    $"{Process.GetCurrentProcess().Id}-{buildConfig}-{runtimeIdentifier}");
+
+                tempFolders.Add(publishOutputTempDir);
+
+                if (!Directory.Exists(publishOutputTempDir))
+                    Directory.CreateDirectory(publishOutputTempDir);
+
+                // Execute dotnet publish
+
+                if (!BuildManager.PublishProjectBlocking(buildConfig, platform,
+                        runtimeIdentifier, publishOutputTempDir))
+                {
+                    throw new InvalidOperationException("Failed to build project.");
+                }
+
+                string soExt = ridOS switch
+                {
+                    OS.DotNetOS.Win or OS.DotNetOS.Win10 => "dll",
+                    OS.DotNetOS.OSX or OS.DotNetOS.iOS => "dylib",
+                    _ => "so"
+                };
+
+                if (!File.Exists(Path.Combine(publishOutputTempDir, $"{GodotSharpDirs.ProjectAssemblyName}.dll"))
+                    // NativeAOT shared library output
+                    && !File.Exists(Path.Combine(publishOutputTempDir, $"{GodotSharpDirs.ProjectAssemblyName}.{soExt}")))
+                {
+                    throw new NotSupportedException(
+                        "Publish succeeded but project assembly not found in the output directory");
+                }
+
+                // Add to the exported project sharead object list.
+
+                foreach (string file in Directory.GetFiles(publishOutputTempDir, "*", SearchOption.AllDirectories))
+                {
+                    if (platform == OS.Platforms.MacOS)
+                    {
+                        AddSharedObject(file, tags: null, Path.Combine("Contents", "Resources", DetermineDataDirNameForProject() + "_" + arch));
```

Comment:
Nitpick but the call to `DetermineDataDirNameForProject` could be done outside the for loop, may also want to use string interpolation instead of concatenation to avoid some string allocations.

```csharp
$"{DetermineDataDirNameForProject()}_{arch}"
```